### PR TITLE
fix(tags): fix inheritance according to cucumber specs

### LINKED ISSUE #196: [feature request] use tags for example tables for scenarios with more example tables
When I have a scenario with more example tables with different tags:
```
Scenario Outline: Test example
  Given Step with <param1>
  When Step with <param2>
  Then Result is <result>

@smoke
Examples:
| param1 | param2 | result |
| foo    | bar    | foobar |
| 1      | 2      | 3      |

@regression
Examples:
| param1 | param2 | result |
| abc    | def    | abcdef |
| z      | y      | z      |
```
I would like to run only those examples that are tagged with specific tag, for example `@smoke`.
This is possible with pure [cucumber-js](https://github.com/cucumber/cucumber-js/blob/master/docs/cli.md#tags), do they use different [tag-expressions](https://github.com/cucumber/cucumber/tree/master/tag-expressions)?
According to cucumber [documentation](https://cucumber.io/docs/cucumber/api/#tags):
Tags can be placed above the following Gherkin elements:
- Feature
- Scenario
- Scenario Outline
- Examples

I tried to run cucumber-js and it worked, but when running cypress with cypress-cucumber-preprocessor, I get an error: `We could not detect any tests in the above file. Write some tests and re-run.`

### PR DESCRIPTION
**Pull request by Sonaryr**
According to cucumber tags should be inherited downwards. See
https://cucumber.io/docs/cucumber/api/#tags

fix #196

### CODE REVIEW

**Comment by Sonaryr** [comment]
So any response on this or not?

**Comment by lgandecki** [comment]
@MateuszNowosad could you please verify and merge if this looks good?

**Comment by MateuszNowosad** [comment]
I will test this today.

**Review by MateuszNowosad — CHANGES_REQUESTED** [review]
From what I can see this follows the existing pattern used in creating tests from scenarios. I wasn't able to break it, everything seems to be working correctly. I've also run commented out tests from BeforeAndAfterSteps and BailOnStepFailure just to be sure. 

I've put forward some suggestions for @Sonaryr to look at before merging. Maybe I will come up with something that might need fixing tomorrow but if not this will be ready for a merge.

File: cypress/integration/BeforeAndAfterSteps.feature
Comment (Sonaryr) [reply cypress/integration/BeforeAndAfterSteps.feature]: Changed

File: cypress/integration/BeforeAndAfterSteps.feature
Comment (Sonaryr) [reply cypress/integration/BeforeAndAfterSteps.feature]: Changed

File: cypress/integration/TagsInheritance.feature
Comment (Sonaryr) [reply cypress/integration/TagsInheritance.feature]: Removed

**Review by MateuszNowosad — COMMENTED** [review]
@Sonaryr Sorry, it took me a few days to re-review this. The current state of this PR is satisfactory however, while I was testing this I noticed some issues. 

While having e.g. test structure like:

```
@all
Feature
    
    @tag1
    Scenario Outline: I order <product> 

        @tag2
        Examples:
        
        @tag3
        Examples:
```

if you were to try:
`./node_modules/.bin/cypress-tags run -e TAGS='@all and @tag2'`

No test would actually run because [cypress-tags.js](https://github.com/TheBrainFamily/cypress-cucumber-preprocessor/blob/b9804ce4a647310320bd4f17b42391f8b46cfe07/cypress-tags.js) file was not changed to accommodate Example tagging. This issue was raised in #489. 

There would have to be some changes made to this file, somewhere near 85-100 line;

```
const featureTags = parsedFeature.feature.tags;
  const featureShouldRun = shouldProceedCurrentStep(featureTags, envTags);
  const taggedScenarioShouldRun = parsedFeature.feature.children.some(
    (section) =>
      section.tags &&
      section.tags.length &&
      shouldProceedCurrentStep(section.tags.concat(featureTags), envTags)
  );
  debug(
    `Feature: ${featurePath}, featureShouldRun: ${featureShouldRun}, taggedScenarioShouldRun: ${taggedScenarioShouldRun}`
  );
  if (featureShouldRun || taggedScenarioShouldRun) {
    featuresToRun.push(featurePath);
  }
});
```
It would probably be something analogous to taggedScenarioShouldRun. 

Let me know if you would be willing to include this as a part of this PR :)

**Comment by rkrisztian** [comment]
> No test would actually run [...]

In my experience, it is still the last Example that runs, regardless of what tags I selected.

The bigger problem is every test that uses step definitions from `tags_implemention_with_env_set.js` is **useless**: this file does not exercise production code at all:

```js
let isPresentInTagsEnv;
const cypressEnvTags = Cypress.env("TAGS");

Given(/'(.+)' is in current TAGS environmental variable/, (envTagsString) => {
  isPresentInTagsEnv = RegExp(envTagsString).test(cypressEnvTags);
});

Then(/this should (not )?run/, (shouldNotRun) => {
  if (typeof cypressEnvTags !== "undefined") {
    expect(!shouldNotRun).to.equal(isPresentInTagsEnv);
  }
});
```

You can't rely on such tests, they will pass all the time given the Gherkin code was written correctly.

Additionally, all scenarios derived from the examples are named the same way:

```none
  Tags Implementation with environmental variables set only at examples level
    ✓ This scenario should run only 1 example (example #1)
    ✓ This scenario should run only 1 example (example #1)
```

A different numbering is important when the scenario name is not parameterized, although ideally we should add the parameter anyway:

```none
	Scenario Outline: This scenario should <result>
```

But the numbering is still off:

```none
  Tags Implementation with environmental variables set only at examples level
    ✓ This scenario should should run (example #1)
    ✓ This scenario should should not run (example #1)
```

**Comment by rkrisztian** [comment]
For the upcoming weeks I have more capacity to work on all the issues with the multiple Examples sections (including this one). So review for my PR in time is appreciated. :)

**Comment by lgandecki** [comment]
@MateuszNowosad do you think you'd find a moment to take a look at this again?

**Comment by MateuszNowosad** [comment]
> @MateuszNowosad do you think you'd find a moment to take a look at this again?

I will try to take a look at this and new @rkrisztian's PR on the weekend.

**Comment by MateuszNowosad** [comment]
@rkrisztian 

I disagree with 

> > No test would actually run [...]
> 
> In my experience, it is still the last Example that runs, regardless of what tags I selected.

If that's the case, that's an unrelated issue. I reevaluated my previous comment and still stand by the requested changes because the current behavior diverges from expectations. The takeaway from my previous comment is that when someone is using TAGS in the command line like in the example `cypress-tags.js` won't actually recognize that there are some tests that need to run in this file at all.

But I partially agree with 

> The bigger problem is every test that uses step definitions from tags_implemention_with_env_set.js is useless

After digging down on `tags_implemention_with_env_set.js` I agree that tests present in this PR don't fulfill their purpose. This is because that file only really makes sense in the context of the original tests that it was written for. This misunderstanding came from the naming. This PR would require some changes to the tests/step definitions. 

As for your newest PR, I've left a separate comment there but it seems to be good.

I don't know if I will be able to dedicate the time this week to fix the aforementioned issues but maybe;
As to not burden the author of this PR we could merge this as if there was nothing wrong with it and then make a new PR that fixes the missing test and improves `cypress-tags.js` since those aren't critical and shouldn't break anything with the current usages. Essential parts of this PR work correctly and it fixes the problem that it was supposed to fix. What would be your take on this situation @lgandecki?

**Comment by rkrisztian** [comment]
@MateuszNowosad , sure, let's merge this PR first. For the problem I was experiencing, I will just recheck it.

**Comment by lgandecki** [comment]
I've merged it as a feat instead of fix, to not introduce possibly a breaking change in a patch release.

File: lib/createTestFromScenario.js
Comment (bobmccracken) [inline lib/createTestFromScenario.js:151]: @Sonaryr @MateuszNowosad @lgandecki This line introduced the regression in #612. This `shouldRun` does not consider `@focus`.

`shouldRun` is previously set [here](https://github.com/TheBrainFamily/cypress-cucumber-preprocessor/blob/master/lib/createTestsFromFeature.js#L31), but this overwrites it.

**Pull request merged**
